### PR TITLE
Changed scheduled jobs to fail loudly if the user is deleted, and added button to transfer ownership

### LINKED ISSUE #8413: ScheduledJobs running on behalf of a deleted user fail silently
### Environment
* Nautobot version (Docker tag too if applicable): 3.0.4
* Python version: 
* Database platform, version:
* Middleware(s):

### Steps to Reproduce
1. Schedule a job
2. Delete the user associated with the ScheduledJob
3. Job fails silently (no JobResult entry either!)

### PR DESCRIPTION
**Pull request by joewesch**
# Closes #8413
# What's Changed
This is an alternative solution to fixing the issue of Scheduled Jobs failing silently when the user is deleted. This PR replaces both #8920 and #8414.

In this solution, to combat the silent failure we now create a failed JobResult object with a single log message to notify the user that the scheduled job was disabled:

Screenshot 2026-05-05 at 2 11 26 PM

I have also added a banner to the detailed view of the scheduled job when the `obj.user` is not found:

Screenshot 2026-05-05 at 2 11 46 PM

In addition to this, I have added an **Assume Ownership** action button that changes the user to the current user and re-enables the scheduled job:

Screenshot 2026-05-05 at 2 11 54 PM

Screenshot 2026-05-05 at 2 12 06 PM

Once the job is re-enabled, it picks back up. Only a single failed job is recorded for each failed scheduled job.

Screenshot 2026-05-05 at 2 12 49 PM

Note: I coded the Assume Ownership button to always be present - not just when the scheduled job is in an errored state. This helps facilitate the transfer of ownership of a scheduled job at any time in the future.

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)

NTC-4974

### CODE REVIEW

File: nautobot/core/celery/schedulers.py [hunk 1]
```diff
@@ -8,16 +8,92 @@
...
+
+    Also clears the in-memory `user_id` so that subsequent `model.save()` calls
+    (e.g. from `_disable`) cannot raise an FK violation when the FK is orphaned.
+    """
+    # If user_id points to a row that no longer exists (FK orphan, e.g. user was deleted via
+    # raw SQL bypassing on_delete=SET_NULL), any save() of this model will raise an
+    # IntegrityError. Patch the DB directly via queryset update, then null in-memory.
+    if model.user_id is not None:
+        ScheduledJob.objects.filter(pk=model.pk).update(user=None)
+        model.user_id = None
```
Comment (glennmatthews) [inline nautobot/core/celery/schedulers.py:52]: Is this a "real" problem? Or just heavy defensive coding?

File: nautobot/core/celery/schedulers.py [hunk 2]
```diff
@@ -8,16 +8,92 @@
...
+        job_result = JobResult.objects.create(
+            name=model.job_model.name,
+            job_model=model.job_model,
+            scheduled_job=model,
+            user=None,
+            task_name=model.job_model.class_path,
+            status=JobResultStatusChoices.STATUS_FAILURE,
+            date_started=timestamp,
+            date_done=timestamp,
+            traceback=err_message,
```
Comment (glennmatthews) [inline nautobot/core/celery/schedulers.py:73]: Do we care about populating the other fields that a "real" run would include, like `celery_kwargs` for example?

File: nautobot/core/celery/schedulers.py [hunk 3]
```diff
@@ -8,16 +8,92 @@
...
+        )
+        log_object = str(model)[:JOB_LOG_MAX_LOG_OBJECT_LENGTH]
+        absolute_url = ""
+        if hasattr(model, "get_absolute_url"):
+            try:
+                absolute_url = model.get_absolute_url()[:JOB_LOG_MAX_ABSOLUTE_URL_LENGTH]
+            except (AttributeError, NotImplementedError):
+                logger.debug("Scheduled job %s does not provide a usable absolute URL", model.name)
```
Comment (glennmatthews) [inline nautobot/core/celery/schedulers.py]: This should never happen, right? Unless someone passes a `model` that isn't a `ScheduledJob`.

File: nautobot/docs/user-guide/platform-functionality/jobs/job-scheduling-and-approvals.md [hunk 4]
```diff
@@ -23,6 +23,19 @@ If the job requires no approval, it will then be added to the queue of scheduled
 
 Once a job has been scheduled, the schedule can be deleted by navigating to `Jobs > Scheduled Jobs`, selecting the schedule name and clicking the `Delete` button. You can also select multiple schedules and click the `Delete Selected` button to delete them all at once.
 
+### Assuming Ownership of a Scheduled Job
+
+Each scheduled job is owned by the user that created it. If that user is removed, the schedule can no longer run and is automatically disabled with a state of `Errored` the next time it would have fired; a failed `JobResult` is also recorded.
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/platform-functionality/jobs/job-scheduling-and-approvals.md]:
```suggestion
Each scheduled job is owned by, and runs as, the user that created it. If that user is removed, the schedule can no longer run and is automatically disabled with a state of `Errored` the next time it would have fired; a failed `JobResult` is also recorded.
```

File: nautobot/extras/tests/test_models.py [hunk 5]
```diff
@@ -3403,6 +3402,165 @@ def test_save_does_not_override_non_active_status_when_disabled(self):
...
+            task="pass_job.TestPassJob",
+            job_model=self.job_model,
+            user=None,
+            interval=JobExecutionType.TYPE_DAILY,
+            start_time=datetime(year=2050, month=1, day=22, hour=17, minute=0, tzinfo=get_default_timezone()),
+            time_zone=get_default_timezone(),
+        )
+        # job_model is required at the DB layer, but the in-memory mutation exercises the
+        # `_record_missing_user_failure` early-return path.
+        scheduled_job.job_model = None
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_models.py:3490]: Another case that feels like maybe we're being overly defensive in the code, if this is something that is already blocked by database validation.

File: nautobot/extras/views.py [hunk 6]
```diff
@@ -3162,6 +3163,17 @@ class ScheduledJobUIViewSet(
...
     table_class = tables.ScheduledJobTable
     action_buttons = ()
+    extra_detail_view_action_buttons = [
+        object_detail.ExtraDetailViewActionButton(
+            action="assume_ownership",
+            label="Assume Ownership",
+            icon="mdi-account-arrow-right",
+            link_name="extras:scheduledjob_assume_ownership",
+            template_path="extras/inc/scheduledjob_assume_ownership_dropdown_item.html",
+            permission_check=lambda user, obj: can_change(user, obj) and user.has_perm("extras.run_job"),
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: Should we also check `obj.user != user` so as to not show the button when you're already the owner?

File: nautobot/core/templates/components/button/post_extradetailviewactionbutton.html
Comment (glennmatthews) [inline nautobot/core/templates/components/button/post_extradetailviewactionbutton.html:1]: Is this just needed because of `post` vs `get`? Surprised we don't have a generic template for that case already if so.

File: nautobot/extras/views.py [hunk 7]
```diff
@@ -3206,6 +3218,27 @@ def get_extra_context(self, request, instance):
...
+        detail=True,
+        methods=["post"],
+        url_path="assume-ownership",
+        url_name="assume_ownership",
+        custom_view_base_action="change",
+        custom_view_additional_permissions=["extras.run_job"],
+    )
+    def assume_ownership(self, request, pk=None):
+        """Reassign this scheduled job's owner to the requesting user and re-enable it."""
+        obj = self.get_object()
```
Comment (glennmatthews) [inline nautobot/extras/views.py:3238]: Do we need an additional check here on `Job.objects.restrict(request.user, "run").filter(obj.job_model).exists()` or similar - so that you can't assume ownership of a *specific* job you don't have permission to run?

File: nautobot/extras/views.py [hunk 8]
```diff
@@ -3206,6 +3218,27 @@ def get_extra_context(self, request, instance):
...
+        if obj.user_id == request.user.id and obj.enabled and obj.state == ScheduledJobStateChoices.ACTIVE:
+            messages.info(request, f"You already own scheduled job '{obj.name}'.")
```
Comment (glennmatthews) [inline nautobot/extras/views.py]: I wonder if this should just be `if obj.user_id == request.user.id`. It feels weird to me for a user to be able to use this endpoint to re-enable a previously disabled job that they already own.

File: nautobot/core/celery/schedulers.py (hunk 1 above)
Comment (joewesch) [reply nautobot/core/celery/schedulers.py:52]: This was a corner case identified while creating and mocking items in tests. Probably very rare in reality.

File: nautobot/core/celery/schedulers.py (hunk 2 above)
Comment (joewesch) [reply nautobot/core/celery/schedulers.py:73]: I was trying to keep this minimal to convey information and that's it. I can certainly add then if you think we need them.

File: nautobot/core/celery/schedulers.py (hunk 3 above)
Comment (joewesch) [reply nautobot/core/celery/schedulers.py]: I believe Claude copied that pattern from here:

https://github.com/nautobot/nautobot/blob/98ffb8e450114485e4398c2fc47173c78346d181/nautobot/extras/models/jobs.py#L1127-L1148

File: nautobot/core/templates/components/button/post_extradetailviewactionbutton.html
Comment (joewesch) [reply nautobot/core/templates/components/button/post_extradetailviewactionbutton.html:1]: Correct. Would you like me to try and change this to a generic template instead?

File: nautobot/core/celery/schedulers.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/core/celery/schedulers.py]: Makes sense there as `obj` could be one of any number of models (though I think we actually have a baseline bug there as in some cases `obj.get_absolute_url()` can exist but *return* an `AttributeError`, that needs to be fixed at some point). In this case since we are reasonably confident that what we have is specifically a `ScheduledJob` instance, I think this is unneeded here.

File: nautobot/core/templates/components/button/post_extradetailviewactionbutton.html
Comment (glennmatthews) [reply nautobot/core/templates/components/button/post_extradetailviewactionbutton.html:1]: Yeah, I'd think it should be some variant of either `nautobot/core/templates/components/button/formbutton.html` or `nautobot/core/templates/components/button/dropdown.html`. Thanks.

File: nautobot/core/celery/schedulers.py (hunk 2 above)
Comment (glennmatthews) [reply nautobot/core/celery/schedulers.py:73]: Probably fine to leave as minimal, maybe add a comment stating what's being intentionally omitted and why.

File: nautobot/core/celery/schedulers.py (hunk 1 above)
Comment (joewesch) [reply nautobot/core/celery/schedulers.py:52]: Sorry, I was wrong. I checked back in my logs and this was the result from an earlier traceback I got in my initial testing from the beat container.

<details>
    <summary>Full Traceback</summary>

```python
celery_beat-1  | [2026-05-04 16:45:00,020: ERROR/MainProcess] Message Error: insert or update on table "extras_scheduledjob" violates foreign key constraint "extras_scheduledjob_user_id_4d8034ff_fk_auth_user_id"
celery_beat-1  | DETAIL:  Key (user_id)=(4ba7a37c-4d74-48ae-a605-8c72a2905bc8) is not present in table "auth_user".
celery_beat-1  |
celery_beat-1  | ['  File "/usr/local/bin/nautobot-server", line 6, in <module>\n    sys.exit(main())\n', '  File "/source/nautobot/core/cli/__init__.py", line 280, in main\n    execute_from_command_line([sys.argv[0], *unparsed_args])\n', '  File "/usr/local/lib/python3.13/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line\n    utility.execute()\n', '  File "/usr/local/lib/python3.13/site-packages/django/core/management/__init__.py", line 436, in execute\n    self.fetch_command(subcommand).run_from_argv(self.argv)\n', '  File "/source/nautobot/core/management/commands/celery.py", line 23, in run_from_argv\n    celery_main.main()\n', '  File "/usr/local/lib/python3.13/site-packages/click/core.py", line 1435, in main\n    rv = self.invoke(ctx)\n', '  File "/usr/local/lib/python3.13/site-packages/click/core.py", line 1902, in invoke\n    return _process_result(sub_ctx.command.invoke(sub_ctx))\n', '  File "/usr/local/lib/python3.13/site-packages/click/core.py", line 1298, in invoke\n    return ctx.invoke(self.callback, **ctx.params)\n', '  File "/usr/local/lib/python3.13/site-packages/click/core.py", line 853, in invoke\n    return callback(*args, **kwargs)\n', '  File "/usr/local/lib/python3.13/site-packages/click/decorators.py", line 34, in new_func\n    return f(get_current_context(), *args, **kwargs)\n', '  File "/usr/local/lib/python3.13/site-packages/celery/bin/base.py", line 158, in caller\n    return f(ctx, *args, **kwargs)\n', '  File "/usr/local/lib/python3.13/site-packages/celery/bin/beat.py", line 72, in beat\n    return beat().run()\n', '  File "/usr/local/lib/python3.13/site-packages/celery/apps/beat.py", line 84, in run\n    self.start_scheduler()\n', '  File "/usr/local/lib/python3.13/site-packages/celery/apps/beat.py", line 113, in start_scheduler\n    service.start()\n', '  File "/usr/local/lib/python3.13/site-packages/celery/beat.py", line 645, in start\n    interval = self.scheduler.tick()\n', '  File "/source/nautobot/core/celery/schedulers.py", line 207, in tick\n    interval = super().tick(*args, **kwargs)\n', '  File "/usr/local/lib/python3.13/site-packages/celery/beat.py", line 355, in tick\n    self.apply_entry(entry, producer=self.producer)\n', '  File "/usr/local/lib/python3.13/site-packages/celery/beat.py", line 285, in apply_entry\n    exc, traceback.format_stack(), exc_info=True)\n']
celery_beat-1  | Traceback (most recent call last):
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/backends/utils.py", line 105, in _execute
celery_beat-1  |     return self.cursor.execute(sql, params)
celery_beat-1  |            ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django_prometheus/db/common.py", line 69, in execute
celery_beat-1  |     return super().execute(*args, **kwargs)
celery_beat-1  |            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
celery_beat-1  | psycopg2.errors.ForeignKeyViolation: insert or update on table "extras_scheduledjob" violates foreign key constraint "extras_scheduledjob_user_id_4d8034ff_fk_auth_user_id"
celery_beat-1  | DETAIL:  Key (user_id)=(4ba7a37c-4d74-48ae-a605-8c72a2905bc8) is not present in table "auth_user".
celery_beat-1  |
celery_beat-1  |
celery_beat-1  | The above exception was the direct cause of the following exception:
celery_beat-1  |
celery_beat-1  | Traceback (most recent call last):
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/celery/beat.py", line 282, in apply_entry
celery_beat-1  |     result = self.apply_async(entry, producer=producer, advance=False)
celery_beat-1  |   File "/source/nautobot/core/celery/schedulers.py", line 185, in apply_async
celery_beat-1  |     entry.model.save()
celery_beat-1  |     ~~~~~~~~~~~~~~~~^^
celery_beat-1  |   File "/source/nautobot/extras/models/jobs.py", line 1602, in save
celery_beat-1  |     super().save(*args, **kwargs)
celery_beat-1  |     ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/base.py", line 902, in save
celery_beat-1  |     self.save_base(
celery_beat-1  |     ~~~~~~~~~~~~~~^
celery_beat-1  |         using=using,
celery_beat-1  |         ^^^^^^^^^^^^
celery_beat-1  |     ...<2 lines>...
celery_beat-1  |         update_fields=update_fields,
celery_beat-1  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_beat-1  |     )
celery_beat-1  |     ^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/base.py", line 1008, in save_base
celery_beat-1  |     updated = self._save_table(
celery_beat-1  |         raw,
celery_beat-1  |     ...<4 lines>...
celery_beat-1  |         update_fields,
celery_beat-1  |     )
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/base.py", line 1138, in _save_table
celery_beat-1  |     updated = self._do_update(
celery_beat-1  |         base_qs, using, pk_val, values, update_fields, forced_update
celery_beat-1  |     )
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/base.py", line 1203, in _do_update
celery_beat-1  |     return filtered._update(values) > 0
celery_beat-1  |            ~~~~~~~~~~~~~~~~^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/query.py", line 1288, in _update
celery_beat-1  |     return query.get_compiler(self.db).execute_sql(ROW_COUNT)
celery_beat-1  |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/sql/compiler.py", line 2060, in execute_sql
celery_beat-1  |     row_count = super().execute_sql(result_type)
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/models/sql/compiler.py", line 1623, in execute_sql
celery_beat-1  |     cursor.execute(sql, params)
celery_beat-1  |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/backends/utils.py", line 122, in execute
celery_beat-1  |     return super().execute(sql, params)
celery_beat-1  |            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/backends/utils.py", line 79, in execute
celery_beat-1  |     return self._execute_with_wrappers(
celery_beat-1  |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~^
celery_beat-1  |         sql, params, many=False, executor=self._execute
celery_beat-1  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_beat-1  |     )
celery_beat-1  |     ^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
celery_beat-1  |     return executor(sql, params, many, context)
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/backends/utils.py", line 100, in _execute
celery_beat-1  |     with self.db.wrap_database_errors:
celery_beat-1  |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/utils.py", line 91, in __exit__
celery_beat-1  |     raise dj_exc_value.with_traceback(traceback) from exc_value
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django/db/backends/utils.py", line 105, in _execute
celery_beat-1  |     return self.cursor.execute(sql, params)
celery_beat-1  |            ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
celery_beat-1  |   File "/usr/local/lib/python3.13/site-packages/django_prometheus/db/common.py", line 69, in execute
celery_beat-1  |     return super().execute(*args, **kwargs)
celery_beat-1  |            ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
celery_beat-1  | django.db.utils.IntegrityError: insert or update on table "extras_scheduledjob" violates foreign key constraint "extras_scheduledjob_user_id_4d8034ff_fk_auth_user_id"
celery_beat-1  | DETAIL:  Key (user_id)=(4ba7a37c-4d74-48ae-a605-8c72a2905bc8) is not present in table "auth_user".
celery_beat-1  |
celery_beat-1  | 16:45:00.040 ERROR   nautobot.core.celery.schedulers schedulers.py                         __init__() :
celery_beat-1  |   Disabling schedule test_b6014555-4549-4a22-8e0f-918de61cef74 with missing user
celery_beat-1  | [2026-05-04 16:45:00,040: ERROR/MainProcess] Disabling schedule test_b6014555-4549-4a22-8e0f-918de61cef74 with missing user
celery_beat-1  | [2026-05-04 16:45:00,050: INFO/MainProcess] DatabaseScheduler: Schedule changed.
```

</details>

File: nautobot/extras/views.py (hunk 6 above)
Comment (joewesch) [reply nautobot/extras/views.py]: Done

File: nautobot/extras/views.py (hunk 7 above)
Comment (joewesch) [reply nautobot/extras/views.py:3238]: Done

File: nautobot/extras/views.py (hunk 8 above)
Comment (joewesch) [reply nautobot/extras/views.py]: Done

File: nautobot/core/celery/schedulers.py (hunk 1 above)
Comment (glennmatthews) [reply nautobot/core/celery/schedulers.py:52]: Interesting! Would love to understand how we got into that state but seems useful to have in any case. Thanks for the update!

**Review by glennmatthews — APPROVED** [review]

File: nautobot/core/celery/schedulers.py (hunk 2 above)
Comment (joewesch) [reply nautobot/core/celery/schedulers.py:73]: Done

File: nautobot/core/templates/components/button/post_extradetailviewactionbutton.html
Comment (joewesch) [reply nautobot/core/templates/components/button/post_extradetailviewactionbutton.html:1]: Done

File: nautobot/core/celery/schedulers.py (hunk 3 above)
Comment (joewesch) [reply nautobot/core/celery/schedulers.py]: Done
